### PR TITLE
Common: Add Random utilities

### PR DESCRIPTION
**Pull request by leoetlino**
This makes it easier to generate random numbers or fill a buffer with random data in a cryptographically secure way.

This also replaces existing usages of RNG functions in the codebase:

* `<random>` is pretty hard to use correctly, and std::random_device does not give enough guarantees about its results (it's implementation-defined, non cryptographically secure and could even be deterministic on some platforms). Doing things correctly is error prone and verbose.

* rand() is terrible and should not be used especially in crypto code.

### CODE REVIEW

File: Source/Core/Common/Random.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+namespace Common::Random
+{
+void Generate(void* buffer, std::size_t size)
+{
+#if defined(_WIN32)
+  ASSERT(RtlGenRandom(buffer, static_cast<ULONG>(size)));
+#else
+  // Unfortunately, we can't rely on arc4random(3) or getrandom(2) being present.
+  // The former requires libbsd; the latter needs a relatively recent glibc (2.25+).
+  File::IOFile random{"/dev/urandom", "rb"};
```
[the remark this answers is not in the document]
Comment (leoetlino) [reply Source/Core/Common/Random.cpp]: Hmm, I hadn't considered the performance aspect. What do you think of a thread local PRNG object that uses mbedtls's [HMAC_DRBG](https://tls.mbed.org/api/hmac__drbg_8h.html)?

**Comment by delroth** [comment]
HMAC_DRBG should be fine.

File: Source/Core/Common/Random.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+class CSPRNG final
+{
+public:
+  CSPRNG()
+  {
+    mbedtls_entropy_init(&m_entropy);
+    mbedtls_hmac_drbg_init(&m_context);
+    const int ret = mbedtls_hmac_drbg_seed(&m_context, mbedtls_md_info_from_type(MBEDTLS_MD_SHA256),
+                                           mbedtls_entropy_func, &m_entropy, nullptr, 0);
+    ASSERT(ret == 0);
```
Comment (riking) [inline Source/Core/Common/Random.cpp]: Is this a run-time assert or debug only?

Comment (leoetlino) [reply Source/Core/Common/Random.cpp]: Runtime. DEBUG_ASSERT is the debug only version.

**Comment by leoetlino** [comment]
@delroth would you mind reviewing this again? I've made the PRNG object change you suggested, but I'm not 100% sure on the thread_local usage here.

**Comment by matthewharvey** [comment]
This is not passing the cmake build step for me with cmake version 3.11.
Posting here because this is the commit that causes this issue.

The error I get from cmake is:
CMake Error at Source/Core/Common/CMakeLists.txt:162 (target_link_libraries):
  The keyword signature for target_link_libraries has already been used with
  the target "traversal_server".  All uses of target_link_libraries with a
  target must be either all-keyword or all-plain.

  The uses of the keyword signature are here:

   * Source/Core/Common/CMakeLists.txt:160 (target_link_libraries)

Want me to just create an issue report at this point?

**Comment by leoetlino** [comment]
Oh whoops :( can you make a PR to add the missing PRIVATE specifiers?

**Comment by matthewharvey** [comment]
Yeah, absolutely. I just got to work, but I'll create one when I get back.

**Comment by matthewharvey** [comment]
Looks like lioncash beat me to it.
#7095 
Thanks!

**Pull request merged**